Use ffmpeg-headless where applicable.

vs49688 opened the pull request:
###### Description of changes
Following on from https://github.com/NixOS/nixpkgs/pull/196263. Now that we have `ffmpeg-headless`, use it!

Specifically, we can now cleanup `ffmpeg` override for `pipewire`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

vs49688 committed:
nixos/no-x-libs: use headless ffmpeg variants

vs49688 commented:
@Atemu from https://github.com/NixOS/nixpkgs/pull/196263#issuecomment-1279970187
> This would also be helpful for SDL bootstrap.

Where is this? I'd like to include it in this PR, but can't find it.

Atemu commented:
@vs49688 I was misremembering. It was pipewire which depended on SDL2.

Atemu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15739,12 +15739,8 @@ with pkgs;
-        pipewireSupport = false;
-      };
-    };
+    # Break the cycle by depending on ffmpeg-headless.
+    ffmpeg = ffmpeg-headless;
```
What does pipewire use ffmpeg for? Just want to make sure we're not giving it an ffmpeg that's missing critical functionality.

vs49688 replied:
It's only used in https://github.com/PipeWire/pipewire/blob/master/spa/plugins/ffmpeg/ffmpeg.c and friends, and just for encoding/decoding, so only `libavcodec`. The `-headless` changes shouldn't affect this (except _maybe_ the change to `libva-minimal`, but I seriously doubt that would hurt).

Atemu replied:
Might be good to add that as a comment here.

Atemu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18578,7 +18574,9 @@ with pkgs;
   };
 
-  ffmpegthumbnailer = callPackage ../development/libraries/ffmpegthumbnailer { };
+  ffmpegthumbnailer = callPackage ../development/libraries/ffmpegthumbnailer {
+    ffmpeg = ffmpeg-headless;
+  };
```
Nit: I typically prefer making the package itself explicitly depend on the package variant; tweaking its inputs.

vs49688 replied:
Done. I've updated `navidrome` and `unpaper` too.

vs49688 committed:
navidrome: depend on ffmpeg-headless explicitly

vs49688 committed:
unpaper: depend on ffmpeg_5-headless explicitly

vs49688 committed:
ffmpegthumbnailer: switch to ffmpeg-headless

vs49688 committed:
pipewire: use ffmpeg-headless

vs49688 committed:
cyanrip: use ffmpeg-headless

vs49688 committed:
openjfx{11,15,17}: use ffmpeg_4-headless

A quick glance at the source shows it only uses libavcodec.

vs49688 commented:
@ofborg build openjdk17

vs49688 commented:
@ofborg build javaPackages.openjfx17

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1367

vs49688 committed:
gst_all_1: use ffmpeg-headless

gst-libav only uses ffmpeg for its (de)coders and (de)muxers. SDL,
pipewire, and whatever is handled separately.

Depend on ffmpeg-headless to reduce closure size.

SuperSandro2000 commented:
How do we know that eg cyanrip is not broken after this?

SuperSandro2000 commented on pkgs/development/libraries/ffmpegthumbnailer/default.nix:
```diff
@@ -33,7 +33,7 @@ stdenv.mkDerivation rec {
         The only dependencies are ffmpeg and libpng.
     ";
-    platforms = platforms.linux;
+    platforms = platforms.all;
```
```suggestion
    platforms = platforms.unix;
```
darwin+linux is unix for meta.platforms.

vs49688 replied:
Ah, cool, I will fix that up.

vs49688 commented:
> How do we know that eg cyanrip is not broken after this?

* I've manually tested `cyanrip`, `unpaper`, `navidrome`, and `ffmpegthumbnailer`.
* `openjfx*` only uses `libavcodec`, which isn't changed in the headless variant.
* I've tested `gst-libav` via `unoconv` -> `libreoffice` -> `gst-libav`.
    - Looking at the `gst-libav` source code shows that it only uses ffmpeg for some codecs and (de)muxers, nothing that would be really affected by the switch to headless.

I would test everything else more thoroughly, but my computer doesn't really have the chops to rebuild 5000+ packages.

vs49688 committed:
ffmpegthumbnailer: allow build on unix platforms

Switching to ffmpeg-headless allows building on Darwin.

vs49688 commented:
> > How do we know that eg cyanrip is not broken after this?
> 
>     * I've manually tested `cyanrip`, `unpaper`, `navidrome`, and `ffmpegthumbnailer`.
> 
>     * `openjfx*` only uses `libavcodec`, which isn't changed in the headless variant.
> 
>     * I've tested `gst-libav` via `unoconv` -> `libreoffice` -> `gst-libav`.
>       
>       * Looking at the `gst-libav` source code shows that it only uses ffmpeg for some codecs and (de)muxers, nothing that would be really affected by the switch to headless.
> 
> 
> I would test everything else more thoroughly, but my computer doesn't really have the chops to rebuild 5000+ packages.

Oh, and I haven't tested this, but Pipewire only uses libavcodec (via an SPA plugin), which isn't affected by the *-headless changes.

SuperSandro2000 commented:
> I would test everything else more thoroughly, but my computer doesn't really have the chops to rebuild 5000+ packages.

Thats why we target staging. If something breaks please follow up on it.

vs49688 commented:
ping on this?
